metrics/server.go

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/http/pprof"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+type Server struct {
+	port     string
+	handlers []Handler
+
+	server *http.Server
+}
+
+// Handler is the representation of an HTTP handler that would be
+// used by the metrics server to expose the metrics
+type Handler struct {
+	Path        string
+	Description string
+	Handler     http.Handler
+}
+
+// NewServer creates a new metrics server.
+// It receives a handler to expose the metrics from a provider. Right
+// now we're just exposing one handler but in the future it'd support
+// a slice of handlers
+// Also, you can activate/deactivate pprof profiles as well setting
+// the pprof argument to true
+func NewServer(port string, handler Handler, pprof bool) *Server {
+	handlers := []Handler{handler}
+	if pprof {
+		handlers = append(handlers, pprofHandlers()...)
+	}
+	return &Server{port: port, handlers: handlers}
+}
+
+// Start starts the metrics server in the provider port
+func (m *Server) Start() {
+	const (
+		defaultHTTPServerReadTimeoutSeconds  = 30
+		defaultHTTPServerWriteTimeoutSeconds = 30
+	)
+
+	router := mux.NewRouter()
+	router.HandleFunc("/", m.handleRoot)
+	for _, handler := range m.handlers {
+		mlog.Debug("Adding metrics handler", mlog.String("path", handler.Path))
+		router.Handle(handler.Path, handler.Handler)
+	}
+
+	m.server = &http.Server{
+		Addr:         fmt.Sprintf(":%s", m.port),
+		Handler:      router,
+		ReadTimeout:  time.Duration(defaultHTTPServerReadTimeoutSeconds) * time.Second,
+		WriteTimeout: time.Duration(defaultHTTPServerWriteTimeoutSeconds) * time.Second,
+	}
+
+	go func() {
+		mlog.Info("Metrics and profiling server started", mlog.String("port", "8067"))
+		if err := m.server.ListenAndServe(); err != nil {
```

Comment:
Probably I'm missing something but could we include these endpoints to mattermod server?

Reply:
Metrics server is a different system that is decoupled from the server and I'm treating it like that. That's why it's outside the server package and acts aside receiving metrics. Imagine we have it like a microservice that receives and exposes metrics from different sources